mvcc/downgrade.go

```diff
@@ -0,0 +1,134 @@
+	} else {
+		plog.Printf("downgrade meta bucket: remove keyes [%s, %s]", string(confStateKeyName), string(termKeyName))
+	}
+	tx.UnsafeDelete(metaBucketName, confStateKeyName)
+	tx.UnsafeDelete(metaBucketName, termKeyName)
+}
+
+// unsafeReadStorageVersion loads storage version from given backend transaction.
+// Populated since v3.6
+func unsafeReadStorageVersion(tx backend.ReadTx) *semver.Version {
+	_, vs := tx.UnsafeRange(metaBucketName, storageVersionKeyName, nil, 0)
+	if len(vs) == 0 {
+		return nil
+	}
+	v, err := semver.NewVersion(string(vs[0]))
+	if err != nil {
+		return nil
+	}
+	return v
+}
+
+// unsafeReadTerm loads term from given transaction.
+// returns 0 if the data are not found.
+// Term is persisted since v3.5.
+func unsafeReadTerm(tx backend.ReadTx) uint64 {
+	_, ts := tx.UnsafeRange(metaBucketName, termKeyName, nil, 0)
+	if len(ts) == 0 {
+		return 0
+	}
+	return binary.BigEndian.Uint64(ts[0])
+}
+
+// unsafeConfStateFromBackend retrieves ConfState from the backend.
+// Returns nil if confState in backend is not persisted (e.g. backend writen by <v3.5).
+func unsafeConfStateFromBackend(lg *zap.Logger, tx backend.ReadTx) *raftpb.ConfState {
+	keys, vals := tx.UnsafeRange(metaBucketName, confStateKeyName, nil, 0)
+	if len(keys) == 0 {
+		return nil
+	}
+
+	if len(keys) != 1 {
+		if lg != nil {
+			lg.Panic(
+				"unexpected number of key: "+string(confStateKeyName)+" when getting cluster version from backend",
+				zap.Int("number-of-key", len(keys)),
+			)
+		} else {
+			plog.Panicf("unexpected number of key: %s when getting cluster version from backend, number-of-key=%d",
+				string(confStateKeyName), len(keys))
+		}
+	}
+	var confState raftpb.ConfState
+	if err := json.Unmarshal(vals[0], &confState); err != nil {
+		if lg != nil {
+			lg.Panic(
+				"cannot unmarshal confState json retrieved from the backend",
+				zap.ByteString("conf-state-json", vals[0]),
+				zap.Error(err),
+			)
+		} else {
+			plog.Panicf("cannot unmarshal confState json retrieved from the backend, err: %s", err)
+		}
+	}
+	return &confState
+}
+
+// UnsafeDetectSchemaVersion returns version of storage schema. Returned value depends on etcd version that created the backend. For
+// * v3.6 and newer will return storage version.
+// * v3.5 will return it's version if it includes all storage fields added in v3.5 (might require a snapshot).
+// * v3.4 will return it's version if it doesn't include all storage fields added in v3.5.
+func UnsafeDetectSchemaVersion(lg *zap.Logger, tx backend.ReadTx) (v semver.Version, err error) {
+	vp := unsafeReadStorageVersion(tx)
+	if vp != nil {
+		return *vp, nil
+	}
+
+	confstate := unsafeConfStateFromBackend(lg, tx)
+	term := unsafeReadTerm(tx)
+	if confstate == nil && term == 0 {
+		// if both confstate and term are missing, assume it's v3.4
+		return version.V3_4, nil
+	} else if confstate == nil {
+		return v, fmt.Errorf("missing confstate information")
+	} else if len(confstate.Voters) == 0 && term == 0 {
```

Comment:
Why distinguish between confstate set and empty? Don't think the code on main branch looks the same. We should provide more context in comments